server/datastore/cached_mysql/cached_mysql.go

```diff
@@ -2,51 +2,109 @@ package cached_mysql
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
+	"reflect"
 	"time"
 
 	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/jinzhu/copier"
 	"github.com/patrickmn/go-cache"
 )
 
+const (
+	appConfigKey                      = "AppConfig:%s"
+	defaultAppConfigExpiration        = 1 * time.Second
+	packsHostKey                      = "Packs:host:%d"
+	defaultPacksExpiration            = 1 * time.Minute
+	scheduledQueriesKey               = "ScheduledQueries:pack:%d"
+	defaultScheduledQueriesExpiration = 1 * time.Minute
+	teamAgentOptionsKey               = "TeamAgentOptions:team:%d"
+	defaultTeamAgentOptionsExpiration = 1 * time.Minute
+)
+
+// Cloner represents any type that can clone itself. Used by types to provide a more efficient clone method.
+type Cloner interface {
+	Clone() (interface{}, error)
+}
+
+func clone(v interface{}) (interface{}, error) {
+	if cloner, ok := v.(Cloner); ok {
+		return cloner.Clone()
+	}
+
+	vv := reflect.ValueOf(v)
+	isPtr := false
+	if vv.Kind() == reflect.Ptr {
+		isPtr = true
+		vv = vv.Elem()
+	}
+
+	clone := reflect.New(vv.Type())
+	err := copier.Copy(clone.Interface(), v)
+	if err != nil {
+		return nil, err
+	}
+
+	if isPtr {
+		return clone.Interface(), nil
+	}
+
+	return clone.Elem().Interface(), nil
+}
+
+// cloneCache wraps the in memory cache with one that clones items before returning them.
+type cloneCache struct {
+	*cache.Cache
+}
+
+func (w *cloneCache) Get(k string) (interface{}, bool) {
+	v, found := w.Cache.Get(k)
+	if !found {
+		return v, false
+	}
+
+	clone, _ := clone(v)
```

Comment:
Shouldn't it return nil, false if there was an error cloning?

Reply:
Fixed